### ISSUE TITLE
Bot using IFrame stopped working suddenly. Throws Operation returned an invalid status code error

### ISSUE THREAD

**Issue by hshahbaz** [microsoft/BotFramework-WebChat#2959]
SDK: v4
platform: .net core 2.2
Channels: Webchat (both directline and Iframe)

I have a chat bot that was working fine till 2 days ago.  I am using web chat with both DirectLine and Iframe implementation covering different scenario. The web page where I have embedded the bot through IFrame is throwing the above error. 

I have customized the webchat by following this:[https://stackoverflow.com/questions/50510099/change-the-css-of-the-qna-bot-embedded-as-webchat](https://stackoverflow.com/questions/50510099/change-the-css-of-the-qna-bot-embedded-as-webchat)

I tried to debug this using ngrok and what I found is puzzling me. When I changed the message endpoint under bot settings in azure portal the bot started sending me the replies through IFrame. When I changed the message endpoint back to original endpoint I started getting the same error. 

Secondly, instead of using my html page as iframe source `<iframe id="webchat" src='Iframe.html'></iframe>`

I changed it to `
       <iframe id="webchat" src='https://webchat.botframework.com/embed/CivicDevBot?s=<secret key>'></iframe>`

the bot started replying. 

Can anyone help me here. I can provide more information as need be

**Comment by hshahbaz** [comment]
**UPDATE:**

Emulator shows me this message:

![image](https://user-images.githubusercontent.com/32860200/75577557-14ec4e80-5a17-11ea-839c-1dafbeb60023.png)

**Comment by hshahbaz** [comment]
anyone?

**Comment by stevengum** [comment]
@stevkan can you provide an update?

**Comment by AngryVirginian** [comment]
@hshahbaz  the LUIS error you see can also be from that the you are using a LUIS trial key and the key may already be exhausted/expired.  IIRC, you can only perform up to 1,900 LUIS operations with the trial key (can't remember the exact number).

**Comment by stevkan** [comment]
@hshahbaz, @AngryVirginian is right that you should look over they type of key you are using. If that is not the source of your issue, can you tell me how you have LUIS integrated into your bot?

**Comment by stevkan** [comment]
Also, I'm a little confused when you mention changing your own HTML page away from `<iframe id="webchat" src='Iframe.html'></iframe>`. In this scenario, how you initiating Web Chat without the key in the "src"? Or is this the [BotFramework-WebChat](https://github.com/microsoft/BotFramework-WebChat) version, but you are passing it in an iframe window still?

**Comment by hshahbaz** [comment]
Hi @stevkan I think the LUIS error is misleading. I am using two channels with my bot. One is DirectLine and other is Webchat.

The reason I am saying that this error is misleading because the DirectLine channel works perfectly and I have already created a new Key but I still get the same error. 

![image](https://user-images.githubusercontent.com/32860200/75802887-55f3a400-5d32-11ea-8e3d-0c6d7ada4510.png)


I am using v3 of webchat please look at the link on my original post to see what i did to change the UI of the webchat and here is the code for my Iframe.html

`<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>CivicBot</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">
    <style>
        html, body {
            height: 100%;
            margin: 0;
            padding: 0;
            overflow: hidden;
            
        }
       <script
    </style>
    <link href="Style/botchat.css" rel="stylesheet" />
    <link href="Style/botchat-fullwindow.css" rel="stylesheet" />
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js" type="text/javascript"></script>
    
</head>
<body>
    <div id="BotChatElement">
        <div class="wc-app" data-reactroot="">
            <div class="wc-chatview-panel">
                <div class="wc-header"></div><div class="wc-message-pane">
                    <div tabindex="0" class="wc-message-groups">
                        <div class="wc-message-group-content" role="list" aria-live="polite">
                            <div class="wc-message-wrapper list" role="listitem" data-activity-id="9WZ7juTNDimJPZ46nIRKv8-o|0000000">
                                <div class="wc-message wc-message-from-bot">
                                    <div class="wc-message-content">
                                        <svg xmlns="http://www.w3.org/2000/svg" class="wc-message-callout"><path class="point-left" d="m 0 6 l 6 6 v -12 Z" /><path class="point-right" d="m 6 6 l -6 6 v -12 Z" /></svg><div>
                                            <div class="wc-list">
                                                <div class="wc-card wc-adaptive-card">
                                                    <div class="ac-container" style="padding: 8px; box-sizing: border-box; background-color: rgba(0, 0, 0, 0);">
                                                        <div class="ac-container" style="padding: 0px; box-sizing: border-box; background-color: rgba(0, 0, 0, 0);">
                                                            <div style="overflow: hidden; display: flex; box-sizing: border-box; justify-content: flex-start;">
                                                                <div class="ac-container" style="padding: 0px; flex:1 1 auto; min-width: 0px; box-sizing: border-box; background-color: rgba(0, 0, 0, 0);">
                                                                    <div style='text-align: left; color: rgb(0, 0, 0); line-height: 17.29px; font-family: "Segoe UI",sans-serif; font-size: 13px; font-weight: 400; box-sizing: border-box; overflow-wrap: break-word;'>
                                                                        <p style="width: 100%; margin-top: 0px; margin-bottom: 0px;">I can help you answer your question and familarize yourself with CIVIC Financial Services.</p>
                                                                    </div><div style="height: 8px;"></div><div style='text-align: left; color: rgb(0, 0, 0); line-height: 17.29px; font-family: "Segoe UI",sans-serif; font-size: 13px; font-weight: 400; box-sizing: border-box; overflow-wrap: break-word;'>
                                                                        <p style="width: 100%; margin-top: 0px; margin-bottom: 0px;">If you want to talk to a Customer Service Agent, just type "I want to talk to a Customer Service Agent".</p>
                                                                    </div>
                                                                </div>
                                                            </div>
                                                        </div><div style="height: 8px;"></div><div><div style="overflow: hidden;"><div style="display: flex; flex-direction: column; align-items: flex-start;"><button class="ac-pushButton" style="flex:0 1 auto; overflow: hidden; white-space: nowrap; text-overflow: ellipsis;" type="button">About CIVIC</button><div style="height: 8px;"></div><button class="ac-pushButton expandable" style="flex:0 1 auto; overflow: hidden; white-space: nowrap; text-overflow: ellipsis;" type="button">FAQs</button></div></div><div style="background-color: rgba(0, 0, 0, 0);"></div></div>
                                                    </div>
                                                </div>
                                            </div>
                                        </div>
                                    </div>
                                </div><div class="wc-message-from wc-message-from-bot"><span>AIDA at 5:08:42 PM</span></div>
                            </div>
                        </div>
                    </div><div class="wc-suggested-actions"></div>
                </div>
                <div class="wc-console"><label tabindex="0" class="wc-upload" aria-label="Upload file" style="visibility:hidden !important;" hidden;><svg xmlns="http://www.w3.org/2000/svg"><path d="M 19.96 4.79 m -2 0 a 2 2 0 0 1 4 0 a 2 2 0 0 1 -4 0 Z M 8.32 4.19 L 2.5 15.53 L 22.45 15.53 L 17.46 8.56 L 14.42 11.18 L 8.32 4.19 Z M 1.04 1 L 1.04 17 L 24.96 17 L 24.96 1 L 1.04 1 Z M 1.03 0 L 24.96 0 C 25.54 0 26 0.45 26 0.99 L 26 17.01 C 26 17.55 25.53 18 24.96 18 L 1.03 18 C 0.46 18 0 17.55 0 17.01 L 0 0.99 C 0 0.45 0.47 0 1.03 0 Z" /></svg><input tabindex="-1" id="wc-upload-input" role="button" type="file" multiple=""></label><div class="wc-textbox"><input class="wc-shellinput" aria-live="polite" aria-label="Type your message..." type="text" placeholder="Type your message..." value=""></div><button tabindex="0" class="wc-send" role="button" aria-label="Send"><svg xmlns="http://www.w3.org/2000/svg"><path d="M 26.79 9.38 A 0.31 0.31 0 0 0 26.79 8.79 L 0.41 0.02 C 0.36 0 0.34 0 0.32 0 C 0.14 0 0 0.13 0 0.29 C 0 0.33 0.01 0.37 0.03 0.41 L 3.44 9.08 L 0.03 17.76 A 0.29 0.29 0 0 0 0.01 17.8 A 0.28 0.28 0 0 0 0.01 17.86 C 0.01 18.02 0.14 18.16 0.3 18.16 A 0.3 0.3 0 0 0 0.41 18.14 L 26.79 9.38 Z M 0.81 0.79 L 24.84 8.79 L 3.98 8.79 L 0.81 0.79 Z M 3.98 9.37 L 24.84 9.37 L 0.81 17.37 L 3.98 9.37 Z" /></svg></button><button tabindex="0" class="wc-mic hidden inactive" role="button" aria-label="Speak"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 58 58" width="28" height="22"><path d="M 44 28 C 43.448 28 43 28.447 43 29 L 43 35 C 43 42.72 36.72 49 29 49 C 21.28 49 15 42.72 15 35 L 15 29 C 15 28.447 14.552 28 14 28 C 13.448 28 13 28.447 13 29 L 13 35 C 13 43.485 19.644 50.429 28 50.949 L 28 56 L 23 56 C 22.448 56 22 56.447 22 57 C 22 57.553 22.448 58 23 58 L 35 58 C 35.552 58 36 57.553 36 57 C 36 56.447 35.552 56 35 56 L 30 56 L 30 50.949 C 38.356 50.429 45 43.484 45 35 L 45 29 C 45 28.447 44.552 28 44 28 Z" /><path id="micFilling" d="M 28.97 44.438 L 28.97 44.438 C 23.773 44.438 19.521 40.033 19.521 34.649 L 19.521 11.156 C 19.521 5.772 23.773 1.368 28.97 1.368 L 28.97 1.368 C 34.166 1.368 38.418 5.772 38.418 11.156 L 38.418 34.649 C 38.418 40.033 34.166 44.438 28.97 44.438 Z" /><path d="M 29 46 C 35.065 46 40 41.065 40 35 L 40 11 C 40 4.935 35.065 0 29 0 C 22.935 0 18 4.935 18 11 L 18 35 C 18 41.065 22.935 46 29 46 Z M 20 11 C 20 6.037 24.038 2 29 2 C 33.962 2 38 6.037 38 11 L 38 35 C 38 39.963 33.962 44 29 44 C 24.038 44 20 39.963 20 35 L 20 11 Z" /></svg></button></div>
            </div>
        </div>
    </div>
    <script src="Style/botchat.js"></script>
    <script>
        $(document).ready(function () {
            $(".wc-header").append("<div class='wc-welcome-text' style='margin-left:34px;'>Hello, I am AIDA.<div>");

        });
        $(document).ready(function () {
            $(".wc-header").append("<div class='wc-header-image'><img src='Style/AIDA_Header_Image.png' style='height:44px; width: 50px; margin-top: -31px;margin-left:-12px;'/><div>");
        });
    </script>
    <script>
        var model = {
            "userId": "<user id>",
            "userName": "Customer",
            "botId": "<Bot ID>",
            "botIconUrl": "//bot-framework.azureedge.net/bot-icons-v1/CivicDevBot_Fk7CehBnv57u7JTG5O6yvBSG9fgFfYgX2tWDYU60W5M98lE.png",
            "botName": "AIDA",
            "secret": "<secret here>",
            "iconUrl": "//bot-framework.azureedge.net/bot-icons-v1/CivicDevBot_Fk7CehBnv57u7JTG5O6yvBSG9fgFfYgX2tWDYU60W5M98lE.png",
            "directLineUrl": "https://webchat.botframework.com/v3/directline",
            "webSocketEnabled": "true",
            "speechTokenEndpoint": "https://api.botframework.com/v3/speechtoken",
            "speechEnabled": false,
            "cssUrl": "/css/webchat-stable/botchat.css",
            "cssFullWindowUrl": "/css/webchat-stable/botchat-fullwindow.css",
            "botChatJavaScriptUrl": "/scripts/webchat-stable/botchat.js",
            "cognitiveServicesJavaScriptUrl": "/scripts/webchat-stable/CognitiveServices.js"
        };
    </script>
    <script>
        var speechOptions;
    </script>
    <script>
        BotChat.App({
            directLine: {
                secret: model.secret,
                token: model.token,
                domain: model.directLineUrl,
                webSocket: true
            },
            user: { id: model.userId, name: model.userName },
            bot: { id: model.botId, name: model.botName },
            resize: 'window',
            speechOptions: speechOptions,
            locale: 'en'
        }, document.getElementById('BotChatElement'));
    </script>
</body>
</html>


`

**Comment by hshahbaz** [comment]
Here is my take on it: I am using version 3 of webchat and Microsoft recently pushed webchat v4 version. Since then mine did stop work. If I take the latest embed code from Azure for webchat my bot starts working. The new webchat uses webchat-es5.js and the old one used webchat.js. I guess the old webchat.js does not work anymore. 

If i use the new version of webchat I cannot modify the UI of the webchat. If I am wrong at this point please direct me on how to change the UI of the new webchat. I have no qualms on moving to the latest version of webchat except that I have to do a lot of UI work. 

here is a screenshot of how my UI looks currently. 

![image](https://user-images.githubusercontent.com/32860200/75804622-23977600-5d35-11ea-8a59-ba71e0d8c009.png)

**Comment by stevkan** [comment]
@hshahbaz, unfortunately there is no way to alter the CSS of any element when using the v4 iframe Web Chat. This is on purpose as the iframe Web Chat is intended to be minimal and out-of-the-box ready. If customizations are desired, then the Direct Line version of Web Chat should be used. It will require some elbow grease to initially setup, but in the long run I think you'll find it easier to work with for your various customizations.

Regarding webchat.js / webchat-es5.js, both can be used for implementing v4 Web Chat. The "es5" flavor includes certain polyfills needed for running Web Chat in older browsers, like IE 11. The **botchat.js / botchat.css** files are what is used for v3 Web Chat (i.e. BotChat).

_Side note: In case you are unaware, the v3 version of Web Chat is deprecated. It would be in your best interest to adopt the v4 Web Chat tools to ensure the best support from the Web Chat team._

**Comment by hshahbaz** [comment]
Thank you @stevkan I have moved to DirectLine
